automate.py

```diff
@@ -164,6 +167,62 @@ def run_important_traces(self):
                 event.content['message'] = content
                 self._tracer.controller.buffers.app.put(event)
 
+    def schedule_id(self, id_):
+        """Keep track of scheduled ids"""
+        if not isinstance(id_, str):
+            raise AttributeError("Invalid id type, for schedule.")
+        self.ids.add(id_)
+        return id_
+
+    def schedule_traces(self, settings_=settings):
+        """Check for invalid arguments from schedule"""
+        if settings_.TRIGGER_SCHEDULE_TRACES:
+            id_ = self.schedule_id('automatic_traces')
+            trigger = settings_.SCHEDULE_TRIGGER
+            kwargs = settings_.SCHEDULE_ARGS
+            if (not isinstance(kwargs, dict) or
+                    not isinstance(trigger, str)):
+                raise AttributeError("Invalid attributes for "
+                                     "job to be scheduled.")
+            trigger_args = {}
+            trigger_args['kwargs'] = {'trigger': trigger}
+            trigger_args['kwargs'].update(kwargs)
+            self.scheduler.add_callable(id_, self.run_traces,
+                                        **trigger_args['kwargs'])
+            return self.scheduler.get_job(id_)
+        return None
+
+    def schedule_important_traces(self, settings_=settings):
+        """Check for invalid important arguments from schedule"""
+        if settings_.TRIGGER_IMPORTANT_CIRCUITS:
+            id_ = self.schedule_id('automatic_important_traces')
+            trigger = settings_.IMPORTANT_CIRCUITS_TRIGGER
+            kwargs = settings_.IMPORTANT_CIRCUITS_ARGS
+            if (not isinstance(kwargs, dict) or
+                    not isinstance(trigger, str)):
+                raise AttributeError("Invalid attributes for "
+                                     "job to be scheduled.")
+            trigger_args = {}
+            trigger_args['kwargs'] = {'trigger': trigger}
+            trigger_args['kwargs'].update(kwargs)
+            return self.scheduler.add_callable(id_,
+                                               self.run_important_traces,
+                                               **trigger_args['kwargs'])
+        return None
+
+    def unschedule_ids(self, id_set=None):
+        """Remove ids to be unschedule"""
+        if id_set is None:
+            id_set = self.ids.copy()
+        while id_set:
+            id_ = id_set.pop()
+            self.ids.remove(id_)
```

Comment:
```suggestion
            self.ids.discard(id_)
```

It's really nice to have the `id_set` arg, however, there's a potential margin to crash if in the future this is called before any id has added or if the application pre-maturely shutdowns, `self.ids.remove` would result in `KeyError`, could you replace this with `discard` to be safer?

Reply:
Yes, you are right. I will change this.